packages/react/src/hooks/useVisibilityLoader.js

```diff
@@ -0,0 +1,95 @@
+import { useCallback, useEffect, useRef, useState } from 'react';
+import warning from 'warning';
+import { merge } from 'lodash-es';
+
+import { browserSupports } from '../utils/componentUtilityFunctions';
+
+const defaultOptions = {
+  /** are there more items available to load */
+  hasMoreToLoad: true,
+  /** is the application already loading new data */
+  isLoading: false,
+  /** a callback that will be triggered when the ref element becomes visible */
+  onVisible: () => {},
+  /* options passed to IntersectionObserver, see: https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API */
+  intersectionObserverOptions: {
+    rootMargin: '100px 0px 100px 0px',
+  },
+};
+
+/**
+ * A simple hook to return a bool telling if the given element is visible on the screen. It can be
+ * used to load a single item or multiple items. If maxToLoad is greater than 1, the first item in
+ * the returned array will be an array of booleans telling how many of the items are to be loaded. The
+ * loadBatchSize prop determines how many new items to mark as visible each time the reference element
+ * is visible on screen. For example, if the loadBatchSize is 20 and you're scrolling down a table,
+ * when the reference row becomes visible it will update the isVisible array to contain 20 new true
+ * booleans.
+ *
+ * @param {*} ref html element to check if it's visible before rendering more items
+ * @param {object} options
+ * @returns Array an array containing [isVisible: bool | arrayOf(bool), moreToLoad: bool]
+ */
+const useVisibilityLoader = (ref, options) => {
+  if (!browserSupports('IntersectionObserver')) {
+    warning(
+      !__DEV__,
+      'The current browser does not support IntersectionObserver. You will need to include a IntersectionObserver polyfill for this component to function properly.'
+    );
+    return [true, false];
+  }
+
+  const { hasMoreToLoad, isLoading, onVisible, intersectionObserverOptions } = merge(
+    {},
+    defaultOptions,
+    options
+  );
+
+  /* eslint-disable react-hooks/rules-of-hooks */
+  const observerRef = useRef(null);
+  const [isVisible, setIsVisible] = useState(false);
+  const intersectionCallback = useCallback(
```

Comment:
This is a very common "optimization" flaw when using hooks. The use of useCallback here is an attempt to make the code more efficient, but in fact, it makes it less efficient.

The passed function to useCallback() is going to get created every time.  Just now with useCallback it accepts the function and array and figures out if it should return the newly created function or a memorized one, wasting CPU cycles since the end result is that the callback function doesn't need to be memoized.

given that the useEffect is monitoring 2 of the 3 props here, this would actually be better if you didn't use useCallback and just put isVisible in the useEffect.  The end result is the same with less CPU cycles.

If you disagree... Read this.  https://kentcdodds.com/blog/usememo-and-usecallback

I'm curious why you had to disable rules-of-hooks.  Was it warning about this very thing?

Reply:
Good catch. I can adjust that.